python/accumulation.py

```diff
@@ -0,0 +1,712 @@
+
+        self.max_frames_per_radar = max_frames_per_radar
+        self.snr_threshold = snr_threshold
+        self.max_frames_without_filter = max_frames_without_filter
+        self.point_filter = (
+            point_filter if max_frames_without_filter < max_frames_per_radar else None
+        )
+        self.radar_filter = radar_filter
+        self.allow_no_extrinsics = allow_no_extrinsics
+
+        self._buffers = dict()
+        self._extrinsics = dict()
+
+    def accumulate(
+        self,
+        radar_position_id: str,
+        points: [[float]],
+        ego_localization_when_received: RigidTransform,
+        radar_extrinsics: RigidTransform = None,
+    ):
+        """Accumulates next radar point cloud.
+
+        Args:
+            radar_position_id (str): Radar position id the point cloud originates from, such as "provizio_radar_front_center".
+            points ([[float]]): A list of radar points to accumulate, each being a list of floats matching the format used in provizio_dds.point_cloud2.
+            ego_localization_when_received (RigidTransform): RigidTransform representing the ego position and orientation in some local Euclidean reference frame, usually ENU (see https://en.wikipedia.org/wiki/Local_tangent_plane_coordinates).
+            radar_extrinsics (RigidTransform, optional): When provided specifies the radar position and orientation relative to the ego coordinate frame. Defaults to None, which stands for keeping previous extrinsics of that radar if any, or radar frame = ego frame otherwise.
+
+        Raises:
+            ValueError: In case no extrinsics were ever provided for this radar while allow_no_extrinsics was specified as False when constructing the PointCloudsAccumulator.
+        """
+        assert isinstance(ego_localization_when_received, RigidTransform)
+        assert radar_extrinsics is None or isinstance(radar_extrinsics, RigidTransform)
+
+        if self.radar_filter and not self.radar_filter(radar_position_id):
+            # Simply ignore this radar
+            return
+
+        if radar_position_id not in self._buffers:
+            self._buffers[radar_position_id] = collections.deque(
+                maxlen=self.max_frames_per_radar
+            )
+
+        buffer = self._buffers[radar_position_id]
+
+        if self.snr_threshold > 0:
+            points = list(
+                filter(
+                    lambda point: (
+                        point[PointCloudsAccumulator._snr_index_in_point]
+                        >= self.snr_threshold
+                    ),
+                    points,
+                )
+            )
+
+        if self.point_filter:
+            if self.max_frames_without_filter <= 0:
+                points = list(filter(self.point_filter, points))
+            elif len(buffer) >= self.max_frames_without_filter:
+                buffer[-self.max_frames_without_filter].points = filter(
```

Comment:
Good catch, fixed.

Reply:
@kevmcapro Turns out, surprisingly, there already was a test: `test_accumulate_snr_and_velocity_filters`, which worked just fine. To tell the truth I've no idea why it still worked without `list()`...

Reply:
I think I might know why it still works. In `get_points_local_frame_relative`, the following line reads each buffered pointcloud. Some of these will be of type `list` and others of type `filter`.

https://github.com/provizio/provizio_dds/blob/860ffafe6a1fd81e8b1865570b3688721230787b/python/accumulation.py#L297C17-L297C51

Both `list` and `filter` can be used as an iterator, therefore it works the same way regardless of if it is a `list` and `filter` because they both support iteration. It would break if you tried to index a `filter` object, because indexing is not supported. I guess this is the kind of thing that can sneak through with a weakly typed language like Python. Maybe more type hints and a strict type checker like Mypy could help prevent this in the future.

Reply:
Done